marlowe-dashboard-client/src/Marlowe/Execution.purs

```diff
@@ -4,48 +4,35 @@ import Prelude
 import Data.Array (fromFoldable)
 import Data.BigInteger (BigInteger, fromInt)
 import Data.Lens (Lens', view)
-import Data.Lens.Iso.Newtype (_Newtype)
 import Data.Lens.Record (prop)
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(..))
 import Data.Symbol (SProxy(..))
-import Marlowe.Semantics (AccountId, Action(..), Bound(..), Case(..), ChoiceId(..), ChosenNum, Contract(..), Input, Observation, Party(..), Payment, Slot(..), SlotInterval(..), State, Timeout, Token(..), TransactionInput(..), TransactionOutput(..), ValueId, _boundValues, _minSlot, computeTransaction, emptyState, evalValue, makeEnvironment)
-
--- Represents a historical step in a contract's life and is what you see on a Step card that is in the past,
--- that is the State as it was before it was executed and the TransactionInput that was applied.
--- We don't bother storing the Contract because it is not needed for displaying a hostorical card but this means
--- we need to store if the step timed out. This is all (possibly premature) optimization to avoid storing the
--- contract many times as it could be quite large
-type ExecutionStep
-  -- FIXME: If the transaction was a timeout, we don't actually have txInput and the state should be the
-  -- one before or the one after.
-  -- If timeout I'll need from the contract the timeout slot
-  -- For the balances is still not clear if we should use the balance before, the balance after or dont display.
-  = { txInput :: TransactionInput
+import Debug.Trace (spy)
+import Marlowe.Semantics (AccountId, Action(..), Bound, Case(..), ChoiceId(..), ChosenNum, Contract(..), Input, Observation, Party, Payment, ReduceResult(..), Slot(..), SlotInterval(..), State, Timeout, Timeouts(..), Token, TransactionInput(..), TransactionOutput(..), ValueId, _boundValues, _minSlot, computeTransaction, emptyState, evalValue, makeEnvironment, reduceContractUntilQuiescent, timeouts)
+
+-- Represents a historical step in a contract's life.
+data ExecutionStep
+  = TransactionStep
+    { txInput :: TransactionInput
+    -- This is the state before the txInput was executed
+    , state :: State
+    }
+  | TimeoutStep
+    { timeoutSlot :: Slot
+    -- This is the state the step had at the begining (before it timed-out)
```

Comment:
`type ExecutionStep = { state :: State, txInput :: Either TransactionInput Slot }`  ?? 🤔

Reply:
I prefer using a custom sum type than either for these cases. I could factor out the `state` in any case, but I'm doing a version of this PR with more changes to the Execution state as this has some problems